### DISCUSSION TITLE
how to include p/invoke needed native library in csproj and be able to publish the project?

### DISCUSSION THREAD

**Discussion by John0King** [dotnet/runtime#65420] [Q&A]
from other articles , that I should use follow csproj to create a nuget 
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <RuntimeIdentifiers>win-x64;win-x86</RuntimeIdentifiers>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <IncludeBuildOutput>false</IncludeBuildOutput>
    <GenerateLibraryLayout>true</GenerateLibraryLayout>
  </PropertyGroup>

  <ItemGroup>
    <!--  there is an  runtimes\win-x64\native\vbar.dll  and runtimes\win-x86\native\vbar.dll  -->
    <Content Include="runtimes\**\*.dll">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Pack>true</Pack>
      <PackagePath>runtimes</PackagePath>
    </Content>
  </ItemGroup>

</Project>

```

the problem is that my other project reference this project  and can not use this native dll,  there no reference point to this `vbar.dll` 

and my app project use follow configuration
```xml
<ProjectReference Include="..\<native folder>\<native folder name>.csproj">
      <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
      <IncludeAssets>all</IncludeAssets>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </ProjectReference>
```

## the error behavior

-  publish ready2run and publish single file  will not copy the vbar.dll from  runtimes(this folder is exists but other native files are copied to the top level folder  and no entry in .deps.json  ) and cause the p/invoke does not work
-  there will be an error if I set `ReferenceOutputAssembly` to `true`

**Comment by John0King** [comment]
I wonder is this a  design flaw  that nuget can reference native binary but project can not !
there should be a `<NativeBinaryReference Include="point-to-.dll" rid="win-x64" />`   and that can be use to generate both the `.deps.json` and nuget layout.   (there is a `<NativeReference />`  I do not know what it can do and how to use it)

**Comment by vitek-karas** [comment]
This is a known limitation right now (you could call it a design flaw). There are multiple issues on this topic in the dotnet/sdk repo - for example: https://github.com/dotnet/sdk/issues/16381.
The short version - there's no good way to include native dependencies in a project reference, the only way which really works is package reference.
